btcec/schnorr/musig2/context.go

```diff
@@ -0,0 +1,280 @@
+// Copyright (c) 2013-2022 The btcsuite developers
+
+package musig2
+
+import (
+	"fmt"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcec/v2/schnorr"
+)
+
+var (
+	// ErrSignerNotInKeySet is returned when a the private key for a signer
+	// isn't included in the set of signing public keys.
+	ErrSignerNotInKeySet = fmt.Errorf("Signing key is not found in key" +
+		" set")
+
+	// ErrAlredyHaveAllNonces is called when RegisterPubNonce is called too
+	// many times for a given signing session.
+	ErrAlredyHaveAllNonces = fmt.Errorf("Already have all nonces")
+
+	// ErrAlredyHaveAllSigs is called when CombineSig is called too many
+	// times for a given signing session.
+	ErrAlredyHaveAllSigs = fmt.Errorf("Already have all sigs")
+
+	// ErrSigningContextReuse is returned if a user attempts to sign using
+	// the same signing context more than once.
+	ErrSigningContextReuse = fmt.Errorf("Nonce already used")
+
+	// ErrFinalSigInvalid is returned when the combined signature turns out
+	// to be invalid.
+	ErrFinalSigInvalid = fmt.Errorf("Final signature is invalid")
+)
+
+// Context is a managed signing context for musig2. It takes care of things
+// like securely generating secret nonces, aggregating keys and nonces, etc.
+type Context struct {
+	// signingKey is the key we'll use for signing.
+	signingKey *btcec.PrivateKey
+
+	// pubKey is our even-y coordinate public  key.
+	pubKey *btcec.PublicKey
+
+	// keySet is the set of all signers.
+	keySet []*btcec.PublicKey
+
+	// combinedKey is the aggregated public key.
+	combinedKey *btcec.PublicKey
+
+	// uniqueKeyIndex is the index of the second unique key in the keySet.
+	// This is used to speed up signing and verification computations.
+	uniqueKeyIndex int
+
+	// keysHash is the hash of all the keys as defined in musgi2.
+	keysHash []byte
+
+	// shouldSort keeps track of if the public keys should be sorted before
+	// any operations.
+	shouldSort bool
+}
+
+// NewContext creates a new signing context with the passed singing key and set
+// of public keys for each of the other signers.
+//
+// NOTE: This struct should be used over the raw Sign API whenever possible.
+func NewContext(signingKey *btcec.PrivateKey,
```

Comment:
> Could we abstract the signing key behind an interface so we don't need to specify it here already? I'm trying to think of a way to support remote signing in a way where the watch-only node could keep track of the signing sessions and would only redirect the raw Sign() call to the signing node

Hmm, I think that's possible. But in the remote signing setting, isn't it the case that the caller already doesn't have access to the private key, and all the session state would actually live entirely in the remote signer? 

I can see how it can be nice to have the signing node be stateless, but given that it's the one that needs to manage the nonce state (which is where the danger is), I'm not sure how we can avoid it having that extra _in memory state_ (assuming all protocols are designed to drop the nonce state if things need to resume of connections be re-established or w/e). 

> Maybe a new SignSchnorr RPC on the signing service would help here?

Yeah this was my original idea, but then I saw the way you sort of made the decision implicitly when signing outputs to figure out if ECDSA vs schnorr should be used. One thing to keep in mind here though, is that the schnorr partial signature *is not* the same as a normal BIP-340 signature. So we'd need a `SignPartialSchnorr` that would only be used in the musig2 case, with something else being used for generic bip 340 signatures outside of the Bitcoin sighash scenario.